pkg/proxy/iptables/proxier.go

```diff
@@ -204,11 +204,8 @@ func CleanupLeftovers(ipt utiliptables.Interface) (encounteredError bool) {
 		encounteredError = true
 	}
 
-	args = []string{"-m", "comment", "--comment", "kubernetes service traffic requiring SNAT", "-m", "mark", "--mark", iptablesMasqueradeMark, "-j", "MASQUERADE"}
-	if err := ipt.DeleteRule(utiliptables.TableNAT, utiliptables.ChainPostrouting, args...); err != nil {
-		glog.Errorf("Error removing pure-iptables proxy rule: %v", err)
-		encounteredError = true
-	}
+	// remove all masquerade rules from the POSTROUTING chain.
+	encounteredError = ensureMasqueradeRuleOnly(ipt, "") || encounteredError
```

Comment:
Most of this effort is dealing with old rules floating around for outdated marks.  A single chain jump is a good option: I haven't done it (yet?) because I wasn't sure I could get it done in time, and even with the chain the code to delete the original mark rule still has to stay around to allow people to upgrade without flushing iptables.

Let me know what you think.